### PR TITLE
Support cycles in transitive closures

### LINKED ISSUE #40: Stack overflow error when calculating the closure for a Model that contains a loop
Can reproduce this with a small smithy model like this:

```smithy
namespace test

use smithy4s.meta#adtMember

union SomeUnion {
    optionA: OptionA,
    optionB: OptionB
}

@adtMember(SomeUnion)
structure OptionA {
    a: String
}

@adtMember(SomeUnion)
structure OptionB {
    b: Integer
}
```

when trying to use `smithytranslate.closure.TransitiveModel.compute` to get the closure from, say, `test#SomeUnion`, I get a Stack Overflow error, which I assume is because of the cycle `SomeUnion -> OptionA -> @adtMember -> SomeUnion`.

I don't understand the whole code but I assume it's because we are not checking already visited shape ids. I could give it a try for the fix but some guidance/thoughts would be helpful.

### PR DESCRIPTION
**Pull request by miguel-vila**
This is to fix #40

### CODE REVIEW

File: modules/transitive/src/closure/IdRefVisitor.scala [hunk 1]
```diff
@@ -13,23 +13,24 @@
...
 import software.amazon.smithy.model.Model
 import software.amazon.smithy.model.traits.Trait
-import smithytranslate.closure.TransitiveModel
 
-final class IdRefVisitor(
+private[closure] final class IdRefVisitor(
     model: Model,
     value: Node,
     captureTraits: Boolean,
-    validateModel: Boolean,
```
Comment (miguel-vila) [inline modules/transitive/src/closure/IdRefVisitor.scala:31]: removed this argument from here and a few other places. Model validation can be done just at the end, when we have traversed the model

File: modules/transitive/src/closure/TransitiveModel.scala [hunk 2]
```diff
@@ -24,16 +25,36 @@ import scala.jdk.CollectionConverters._
...
+      validateModel = validateModel,
+      captureMetadata = captureMetadata
+    )
+  }
+
+  private[closure] def computeWithVisited(
+      model: Model,
+      entryPoints: List[ShapeId],
+      captureTraits: Boolean,
+      visitedShapes: mutable.Set[Shape]
```
Comment (miguel-vila) [inline modules/transitive/src/closure/TransitiveModel.scala:56]: this is one of the main additions: keeping track of visited shapes

File: modules/transitive/tests/src/ClosureSpec.scala [hunk 3]
```diff
@@ -212,4 +212,48 @@ class ClosureSpec extends munit.FunSuite {
     )
     assert(result.isSuccess)
   }
+
+  test(
```
Comment (miguel-vila) [inline modules/transitive/tests/src/ClosureSpec.scala:216]: these are 2 kinds of cycles situations I can think of:
- recursive definition
- cyclic reference through traits (this is what made us discover this issue)

File: modules/transitive/src/closure/IdRefVisitor.scala (hunk 1 above)
Comment (daddykotex) [reply modules/transitive/src/closure/IdRefVisitor.scala:31]: absolutely

**Comment by daddykotex** [comment]
Awesome work, this is what I had in mind. I was writing a detailed post to help you onboard but wow, no need for that. I'll review that in detail tomorrow, but it looks great.

File: modules/transitive/src/closure/TransitiveModel.scala (hunk 2 above)
Comment (daddykotex) [reply modules/transitive/src/closure/TransitiveModel.scala:56]: This is fine by me

File: modules/transitive/src/closure/TransitiveModel.scala [hunk 4]
```diff
@@ -53,25 +74,57 @@ object TransitiveModel {
         .map(_.asScala.toSet)
         .fold[Set[Shape]](Set.empty)(_ ++ _)
 
-    val allShapes = (closure ++ entrypointTraits)
+    val allShapes =
+      (closure ++ entrypointTraits).filter(s => !visitedShapes.contains(s))
+    entryPointsShapes.foreach(visitedShapes.add)
```
Comment (daddykotex) [inline modules/transitive/src/closure/TransitiveModel.scala:79]: I was kind of expecting we'd do `allShapes.foreach(visitedShapes.add)`

File: modules/transitive/src/closure/IdRefVisitor.scala (hunk 1 above)
Comment (kubukoz) [reply modules/transitive/src/closure/IdRefVisitor.scala:31]: This is most likely binary incompatible, as you're modifying a previously public class

Comment (daddykotex) [reply modules/transitive/src/closure/IdRefVisitor.scala:31]: I think so too, and we're making private
But at the stage of smithy-translate I don't think we should mind

Comment (kubukoz) [reply modules/transitive/src/closure/IdRefVisitor.scala:31]: actually, with the repackaging any form of BC goes out of the window anyway 😅

File: modules/transitive/src/closure/IdRefVisitor.scala [hunk 5]
```diff
@@ -43,11 +44,17 @@ final class IdRefVisitor(
           .getElements()
```
Comment (kubukoz) [inline modules/transitive/src/closure/IdRefVisitor.scala:44]: not really related to the PR but should `mapShape` be using `visitSeqShape`?

Comment (daddykotex) [reply modules/transitive/src/closure/IdRefVisitor.scala:44]: not only that but we should probably call it for `visitSeqShape(shape.getKey())` ?

**Comment by daddykotex** [comment]
Can I push a commit on top your work @miguel-vila  ?

File: modules/transitive/src/closure/TransitiveModel.scala (hunk 4 above)
Comment (kubukoz) [reply modules/transitive/src/closure/TransitiveModel.scala:79]: also, probably more efficient to do `visitedShapes ++= thing`

File: modules/transitive/src/closure/IdRefVisitor.scala (hunk 5 above)
Comment (kubukoz) [reply modules/transitive/src/closure/IdRefVisitor.scala:44]: that's never going to match `asArrayNode` though, right?

**Comment by miguel-vila** [comment]
@daddykotex feel free to push anything

File: modules/transitive/src/closure/IdRefVisitor.scala [hunk 6]
```diff
@@ -157,21 +168,20 @@ object IdRefVisitor {
       captureTraits: Boolean,
       shapeId: ShapeId,
       trt: Trait,
-      validateModel: Boolean,
-      includeStartingTrait: Boolean = false
```
[the remark this answers is not in the document]
Comment (daddykotex) [reply modules/transitive/src/closure/IdRefVisitor.scala:161]: I was discussing this at $work. I think `includeStartingTrait` is always false.
It is at least in this project, and all of our internal usage.

I don't think it's worth keeping as we stabilize this project

**Comment by daddykotex** [comment]
> @daddykotex feel free to push anything

* Updated with main
* Include a change to use ShapeVisitor.default
* Use expectShape rather than wrapping in option/list

File: modules/transitive/src/closure/IdRefVisitor.scala (hunk 5 above)
Comment (kubukoz) [reply modules/transitive/src/closure/IdRefVisitor.scala:42]: I'm like 90% sure this should be `asObjectNode` for map shapes - both for the key and value. I'm not sure how to reproduce it though.

File: modules/transitive/src/closure/TransitiveModel.scala (hunk 4 above)
Comment (miguel-vila) [reply modules/transitive/src/closure/TransitiveModel.scala]: addressed this at [b0eb2c8](https://github.com/disneystreaming/smithy-translate/pull/42/commits/b0eb2c862445b86ba4c38c236886e0664cab7d04)

File: modules/transitive/src/closure/IdRefVisitor.scala (hunk 5 above)
Comment (daddykotex) [reply modules/transitive/src/closure/IdRefVisitor.scala:42]: Let me have a look, I'll do a final check while I'm there and merge afterward

Comment (daddykotex) [reply modules/transitive/src/closure/IdRefVisitor.scala:42]:
```
=======
shape: (map: `sauce#myTraitMap`)
asArrayNode: Optional.empty
asObjectNode: Optional[software.amazon.smithy.model.node.ObjectNode@b59fc9ec]
asBooleanNode: Optional.empty
asNullNode: Optional.empty
asStringNode: Optional.empty
asNumberNode: Optional.empty
```

Comment (daddykotex) [reply modules/transitive/src/closure/IdRefVisitor.scala:42]: so you're right

**Review by daddykotex — APPROVED** [review]
Thanks for the PR, this is amazing work!
I'll release it in a few hours, mostly under 0.2.0 because we're breaking a few things (moving stuff internally)

**Pull request merged**